### PR TITLE
Development

### PR DESCRIPTION
**Pull request by egtoneyUK**
Stable Build of Development. Everyone look over this and make changes to the development branch or make comments so that we can iron out any mistakes.

### CODE REVIEW

File: SerialProg/ArduinoSerialIO.py [hunk 1]
```diff
@@ -1,51 +1,197 @@
 from ThreadedSerialIO import ThreadedSerialIO
 from datetime import datetime
-import re, time
+import re, time, os
+import sqlite3
```
Comment (egtoneyUK) [inline SerialProg/ArduinoSerialIO.py]: Is this being used?

File: SerialProg/ArduinoSerialIO.py [hunk 2]
```diff
@@ -1,51 +1,197 @@
...
+			match = ArduinoSerialIO.accellerometerPattern.search(self.temp_data)
+			if match:
+				self.temp_data = self.temp_data.replace(match.group(), "")
+				self.parseAccel(match.group())
+
+			#keep only last 100 characters in buffer
+			if len(self.temp_data) >=100:
+				self.temp_data = self.temp_data[len(self.temp_data)-100:]
+
+			time.sleep(0.2)
```
Comment (egtoneyUK) [inline SerialProg/ArduinoSerialIO.py]: We are logging at 5hz so we may want to speed this up to some value that is faster than 5hz

File: SerialProg/BatterySerialIO.py [hunk 3]
```diff
@@ -1,12 +1,16 @@
 from ThreadedSerialIO import ThreadedSerialIO
 from datetime import datetime
-import re, time
+import re, time, os
 import MySQLdb
```
Comment (egtoneyUK) [inline SerialProg/BatterySerialIO.py]: Is this being used?

File: SerialProg/Main.py [hunk 4]
```diff
@@ -3,21 +3,16 @@
...
 from datetime import datetime as dt
 from ThreadedSerialIO import ThreadedSerialIO
-from SerialIdentifier import SerialIdentifier
 from MotorSerialIO import MotorSerialIO
 from BatterySerialIO import BatterySerialIO
 from ArduinoSerialIO import ArduinoSerialIO
 from TelemetrySerialIO import TelemetrySerialIO
+from MpptSerialIO import MpptSerialIO
 from datetime import datetime
+import sqlite3
```
Comment (egtoneyUK) [inline SerialProg/Main.py]: Is this import being used?

File: SerialProg/Main.py [hunk 5]
```diff
@@ -200,52 +176,52 @@ def stop():
...
+		elif search_type == Main.MOTOR:
+			MotorSerialIO.searchForDevices( Main.serial_ports, port_names )
+			time.sleep(Main.OPENING_DELAY)
+		elif search_type == Main.TELEMETRY:
+			TelemetrySerialIO.searchForDevices( Main.serial_ports, port_names )
+			time.sleep(Main.OPENING_DELAY)
+		elif search_type == Main.BATTERY:
+			BatterySerialIO.searchForDevices( Main.serial_ports, port_names )
+			time.sleep(Main.OPENING_DELAY)
+		elif search_type == Main.MPPT:
```
Comment (egtoneyUK) [inline SerialProg/Main.py]: This, Main.MPPT, constant isn't defined

File: SerialProg/Main.py [hunk 6]
```diff
@@ -200,52 +176,52 @@ def stop():
...
+			time.sleep(Main.OPENING_DELAY)
+
+		for p in Main.serial_ports:
+			if( "MOT" in p.port ):
+				Main.motor = p
+			elif( "ARD" in p.port ):
+				Main.arduino = p
+			elif( "TEL" in p.port ):
+				Main.telemetry = p	
+			elif( "MPPT" in p.port ):
```
Comment (egtoneyUK) [inline SerialProg/Main.py]: This elif isn't necessary

File: SerialProg/MotorSerialIO.py [hunk 7]
```diff
@@ -76,14 +86,9 @@ def getStartMilage(self):
...
-			if(len(DB_buffer) == 0):
-				query = "INSERT INTO dbo.bat_batman (id, time, current, voltage, vehicle_velocity, motor_velocity, phase_a_current, phase_b_current, amp_hours, watt_sec, odometer) VALUES "
-			
-			query += "(NULL, NOW(), '" + str(self.bus_current) + "', '" + str(self.bus_voltage) "', " + str(self.vehicle_velocity) + "', " /
-				     + str(self.motor_velocity) + "', " + str(self.phase_a_current) + "', " + str(self.phase_b_current) + "', " /
-				     + str(self.amp_hours) + "', " + str(self.watt_sec) + "', " + str(self.odometer - self.start_milage) + "), " 
-			
-			self.DB_buffer.put(query)
+			current_time = datetime.now()
+			self.insert_query.put("(NULL, '"+(current_time.strftime('%Y-%m-%d %H:%M:%S'))+"', '" + str(self.bus_current) + "', '" + str(self.bus_voltage) + "', '" + str(self.vehicle_velocity) + "', '" + str(self.motor_velocity) + "', '" + str(self.phase_a_current) + "', '" + str(self.phase_b_current) + "', '" + str(self.amp_hours) + "', '" + str(self.calc_watt_sec) + "', '" + str(self.odometer - self.start_milage) + "'), ")
```
Comment (egtoneyUK) [inline SerialProg/MotorSerialIO.py]: Should add first digit of milliseconds onto this for more precision in the MySQL database. This is also true for all other device classes.

File: SerialProg/MpptSerialIO.py [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+				if len(serial_port) > 0 and serial_port not in curr_port_names:
+					new_port = MpptSerialIO( serial_port, "" )
+					list_of_ports.append( new_port )
+
+	def __init__(self, t_port, t_data):
+		super(MpptSerialIO, self).__init__( t_port, 115200, MpptSerialIO.MESSAGE_SIZE, True, t_data, False)
+
+		# Values at the beginning of the network packages
+		self.address
+		self.dataLength
```
Comment (egtoneyUK) [inline SerialProg/MpptSerialIO.py]: @weiliansong Try to use the same variable naming scheme that we use in other files. Here are examples
variables) variable_name_here
functions) functionNameHere()

File: SerialProg/MpptSerialIO.py [hunk 9]
```diff
@@ -0,0 +1,196 @@
+from ThreadedSerialIO import ThreadedSerialIO
+from datetime import datetime
+import re, time, os
+import MySQLdb
+import binascii
+
+class MpptSerialIO(ThreadedSerialIO):
+
+	mpptPattern = re.compile('(t([0-9A-F])*)')
```
Comment (egtoneyUK) [inline SerialProg/MpptSerialIO.py]: All of these Regex pattern variables need to be typed like constants. This needs to be fixed across all device classes.
ex) VARIABLE_CONSTANT

File: SerialProg/TelemetrySerialIO.py [hunk 10]
```diff
@@ -2,25 +2,32 @@
 from BatterySerialIO import BatterySerialIO
 from MotorSerialIO import MotorSerialIO
 from datetime import datetime
-import re, time, threading, subprocess
+import re, time, threading, subprocess, os
+import sqlite3
```
Comment (egtoneyUK) [inline SerialProg/TelemetrySerialIO.py]: Is this import being used?

File: SerialProg/ThreadedSerialIO.py [hunk 11]
```diff
@@ -1,5 +1,6 @@
-import serial, time, io, re, os, queue
+import serial, time, io, re, os, queue, sys
 from threading import Thread
+import sqlite3
```
Comment (egtoneyUK) [inline SerialProg/ThreadedSerialIO.py]: Is this import being used?

File: setup.sh [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+
+READ_RED='\e[0;31m'
+READ_BLUE='\e[0;34m'
+READ_GREEN='\e[0;32m'
+READ_NC='\e[0m' # No Color
+
+# Check to make sure that it has been given sudo privlages
+if [ "$(whoami)" = "root" ] ; then
+	# Usage notice
+	read -p $'\e[0;32m[setup.sh]>\e[0;34m During the installation process you will be required to create a mysql account and to type in your password a few times. Please keep in mind that it does not matter what you make this password. Just make sure that you make it something that it easy for you to remember. Press [Enter] to continue.'
```
Comment (egtoneyUK) [inline setup.sh]: Change from awful dark blue to light blue

Comment (crba233) [inline setup.sh]: About time. :)

File: SerialProg/ThreadedSerialIO.py (hunk 11 above)
Comment (ArthurSilveira) [inline SerialProg/ThreadedSerialIO.py]: Done.

File: setup.sh (hunk 12 above)
Comment (ArthurSilveira) [inline setup.sh]: done xD

File: SerialProg/TelemetrySerialIO.py (hunk 10 above)
Comment (ArthurSilveira) [inline SerialProg/TelemetrySerialIO.py]: Done

**Pull request merged**